drivers/serial/uart_s32_linflexd.c

```diff
@@ -0,0 +1,542 @@
+}
+
+static void uart_s32_irq_err_enable(const struct device *dev)
+{
+	const struct uart_s32_config *config = dev->config;
+	uint32_t linflexd_ier;
+
+	linflexd_ier = sys_read32(POINTER_TO_UINT(&config->base->LINIER));
+
+	/* Enable frame error interrupt and buffer overrun error interrupt */
+	linflexd_ier |= (LINFLEXD_LINIER_FEIE_MASK | LINFLEXD_LINIER_BOIE_MASK);
+
+	sys_write32(linflexd_ier, POINTER_TO_UINT(&config->base->LINIER));
+}
+
+static void uart_s32_irq_err_disable(const struct device *dev)
+{
+	const struct uart_s32_config *config = dev->config;
+	uint32_t linflexd_ier;
+
+	linflexd_ier = sys_read32(POINTER_TO_UINT(&config->base->LINIER));
+
+	/* Disable frame error interrupt and buffer overrun error interrupt */
+	linflexd_ier &= ~(LINFLEXD_LINIER_FEIE_MASK | LINFLEXD_LINIER_BOIE_MASK);
+
+	sys_write32(linflexd_ier, POINTER_TO_UINT(&config->base->LINIER));
+}
+
+static int uart_s32_irq_is_pending(const struct device *dev)
+{
+	return (uart_s32_irq_tx_ready(dev)) || (uart_s32_irq_rx_ready(dev));
+}
+
+static int uart_s32_irq_update(const struct device *dev)
+{
+	return 1;
+}
+
+static void uart_s32_irq_callback_set(const struct device *dev,
+		uart_irq_callback_user_data_t cb,
+		void *cb_data)
+{
+	struct uart_s32_data *data = dev->data;
+
+	data->callback = cb;
+	data->cb_data = cb_data;
+}
+
+/**
+ * @brief Interrupt service routine.
+ *
+ * This simply calls the callback function, if one exists.
+ *
+ * Note: s32 UART Tx interrupts when ready to send; Rx interrupts when char
+ * received.
+ *
+ * @param arg Argument to ISR.
+ *
+ * @return N/A
+ */
+
+void uart_s32_isr(const struct device *dev)
+{
+	struct uart_s32_data *data = dev->data;
+
+	if (data->callback) {
+		data->callback(dev, data->cb_data);
+	}
+}
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+/**
+ * @brief Initialize UART channel
+ *
+ * This routine is called to reset the chip in a quiescent state.
+ * It is assumed that this function is called only once per UART.
+ *
+ * @param dev UART device struct
+ *
+ * @return 0
+ */
+static int uart_s32_init(const struct device *dev)
+{
+	const struct uart_s32_config *config = dev->config;
+	struct uart_s32_data *data = dev->data;
+	static uint8_t state_idx;
+	uint8_t key;
+	int err;
+
+	err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (err < 0) {
+		return err;
+	}
+
+	key = irq_lock();
+
+	/* Initialize UART with default configuration */
+	data->hw_cfg.BaudRate = 115200;
+	data->hw_cfg.BaudRateMantissa = 26U;
+	data->hw_cfg.BaudRateFractionalDivisor = 1U;
+	data->hw_cfg.ParityCheck = false;
+	data->hw_cfg.ParityType = LINFLEXD_UART_IP_PARITY_EVEN;
+	data->hw_cfg.StopBitsCount = LINFLEXD_UART_IP_ONE_STOP_BIT;
+	data->hw_cfg.WordLength = LINFLEXD_UART_IP_8_BITS;
+	data->hw_cfg.TransferType = LINFLEXD_UART_IP_USING_INTERRUPTS;
+	data->hw_cfg.Callback = s32_uart_callback;
+	data->hw_cfg.CallbackParam = NULL;
+	data->hw_cfg.StateStruct = &Linflexd_Uart_Ip_apStateStructure[state_idx++];
```

Comment:
is state_idx correct?  what happens if I have uart1 and uart5 enabled?

Reply:
it's correct, `Linflexd_Uart_Ip_apStateStructure` length is determined based on the number of instances enabled, and not the total number of instances available. Pls see https://github.com/zephyrproject-rtos/hal_nxp/pull/182/commits/eb424444215a27cbf20053595068bd25d54991db
So if `uart1` and `uart5` are enabled, the length of the array is 2, `uart1` will use index 0 and `uart2` index 1.